### DISCUSSION TITLE
How to type hint a decorator that injects a parameter and have the return type validated correctly

### DISCUSSION THREAD

**Discussion by crimsonknave** [microsoft/pyright#7277] [Q&A] [answered]
I'm building a decorator that injects a parameter into the decorated function. This lets the decorated function have that parameter as an argument, but the caller does not pass it, the decorator does. I've tried so many different ways to get this accurately typed, but after hours of trying, I have not come up with solution where pyright will complain when a decorated function is assigned to something incompatible with it's return type, but also not complain that the injected parameter is missing from the call.

It feels like this should be doable and I must be missing something. But, no example I could find for this sort of situation has done both.

Here's a simplified version my decorator that illustrates the problem. The real decorator uses `makefun.wraps` for dynamic signature matching and injects the new parameter via the kwargs, which works, but I acknowledge might not be something I can type fully. If I have to require the injected parameter to be in a specific order in order to get typing properly, that's fine. (I also assume I'll have to have overrides/unions to account for instance/class methods vs bare methods and the self/cls param.)

The thing that confuses me the most is why changing the return type to be a `TypeVar` has any impact on what parameters are expected.

```python
import contextlib
from functools import wraps
from typing import (
    Any,
    Callable,
    Concatenate,
    ParamSpec,
    TypeVar,
)

P = ParamSpec("P")
Return1 = TypeVar("Return1")
Return2 = Any
ToDecorate1 = Callable[Concatenate[str, P], Return1]
ToDecorate2 = Callable[Concatenate[str, P], Return2]
Decorated1 = Callable[P, Return1]
Decorated2 = Callable[P, Return2]
Decorator1 = Callable[[ToDecorate1], Decorated1]
Decorator2 = Callable[[ToDecorate2], Decorated2]


def test_decorator1() -> Decorator1:
    def decorate(function: ToDecorate1) -> Decorated1:
        @wraps(function)
        def wrapper(*args: P.args, **kwargs: P.kwargs):
            return function("inserted1", *args, **kwargs)

        return wrapper

    return decorate


def test_decorator2() -> Decorator2:
    def decorate(function: ToDecorate2) -> Decorated2:
        @wraps(function)
        def wrapper(*args: P.args, **kwargs: P.kwargs):
            return function("inserted2", *args, **kwargs)

        return wrapper

    return decorate


@test_decorator1()
def go1(removed_param: str) -> int:
    print(removed_param)
    return 1


@test_decorator2()
def go2(removed_param: str) -> int:
    print(removed_param)
    return 2


@test_decorator2()
def go2str(removed_param: str) -> str:
    print(removed_param)
    return "2"


# This works correctly, but pyright thinks it's missing a parameter
missing1: str = go1()

# This does not work, but does give the correct type error for the return type not matching
with contextlib.suppress(TypeError):
    wrong_type1: str = go1("")

# These work, but do not give an error for the return type not matching
int2: str = go2()
str2: int = go2str()
```

**Comment by erictraut** [comment, marked as answer]
You have defined several generic type aliases. For example, `ToDecorate1` is a generic type alias that accepts a single type argument corresponding to type parameter `P`. However, when you later make use of these type aliases, you are not providing any type arguments. That means a type checker will default to `Any` for regular TypeVars and to `...` for ParamSpecs. That's probably not your intent here. If you enable the `reportMissingTypeArgument` check, pyright will point out these (probably unintended) cases where you've omitted the type argument.

Here's a corrected code sample in [pyright playground](https://pyright-play.net/?enableExperimentalFeatures=true&code=MQAgDgngTglg5gCwC4C4RQKZgPZSQWRgGciYA7OAFQjAwEEo4BXAWwzKQF4komMBYAFBCAZlGwsQIpmQDGSbNgA2REDBY48IAO5QAhmCKjxkpDXJw1G3EhB0yEADQgAwnqVK9AIyUZnL7Dk9JHZgvxAABT19FgBlWllnaloANWihIQiQTkjovTiEgAoAIgjigEohACUMJCYoMgBGbJBkjDSoEpq6hsaK6tr6sgAmFvsIIUpsABEMWVww5py3D29fAG0AoJCyMPWiHmcIgF1nbqHG48mZuYWQ0eX3Tx8MTcDZMN2Q-cPI05Bzg1hldBLN5voQgATJauJ5rV4RM6DXogsF3DCQh6w1YvdaIgHIkao24Q3AwlbPDbrKZoiEYRp4pE9JrHf60sLQxkE5mXYng4K4LEU%2BHrak3fn3PGs5zsqHDKUgjKCSEYEQgEIHAD6KolZMK5RAAFoAHwgWVkrmAlkoIQgO0gFVqnXowrSORIGCBNA0kmLS2Ey4Gk1m31Qhn4q2XG2Ce2xkAAAV0BiIrpk8k9ZEqMbjdsdOn0YFonQAVNE4EQ0BEAHRlojOYvFgDW2lrlarzdr5WjOZzmGZUjTHsCJXIRAweAxfXrtfrTZbjCIWdtPb7Q3zBiLStjq4aDtDAmEh7zGqQ2tDgv1RtN5qg8pO3fteeddNT7oz3vF6LvxyD1-3mKlB8c0TAsUzddNAizHs9zVJNC3HQpSwXNsZxABsO2QyJ23ncsu2XaC7R3MgBzfYdilHcc5WKacF1nDDcK3FdCXXeCoEYwjmOfMIlSEeMTzPXUoEafUhDzOBsGEzAWGwAA3DFNTAPIWDQA4oF-NQOCA8BYA4QopNk%2BTFJiKD7SIkBGh4wQ%2BIwLUuIUW8ROVVUQHE4Y9IwaS5MhBSlJUnh1PIVB8O0wL3M8wylJMjj%2B2GSzrNs88HKzMTsGGVSwoM7yjPyPy1KvEBVK0sAdKQDKvJ84zgrM4phmKSyZMaNBApacThKzGThlylrUscmSAGYmo4bq0p4RygA).

**Reply by crimsonknave** [reply]
Thanks! That's very helpful, one more question. Is there a better way to have the type hints know how to deal with both instance and non instance methods? (Or support both modifying the signature or not?) Right now I've got a bajillion custom types defined as well as overloads for all the permutations of parameter options. It works, but it's quite verbose and feels like it will be/get brittle. 

The decorator is defined in a class where it stores some config, if that matters. I did initially run into an issue where I had made the class generic and passed in the types `class Foo(Generic[S, P, R]):`, but that broke all typing of returns. Best I could tell the fact that different decorated functions could return different types meant that they were all using the same `R`. Everything worked when I removed the `Generic` from the class.

```python
P = ParamSpec("P")
R = TypeVar("R")
S = TypeVar("S")

SelfLoggerAndParams = Callable[Concatenate[S, "AnnotatedAdapter", P], R]
LoggerAndParams = Callable[Concatenate["AnnotatedAdapter", P], R]
SelfAndParams = Callable[Concatenate[S, P], R]
ParamsOnly = Callable[P, R]
SelfAndLogger = Callable[[S, "AnnotatedAdapter"], R]
LoggerOnly = Callable[["AnnotatedAdapter"], R]
SelfOnly = Callable[[S], R]
Empty = Callable[[], R]

NoInjectionSelf = Callable[[SelfAndParams[S, P, R]], SelfAndParams[S, P, R]]
NoInjectionBare = Callable[[ParamsOnly[P, R]], ParamsOnly[P, R]]
InjectionSelf = Callable[[SelfLoggerAndParams[S, P, R]], SelfAndParams[S, P, R]]
InjectionSelfProvide = Callable[
    [SelfLoggerAndParams[S, P, R]], SelfLoggerAndParams[S, P, R]
]
InjectionBare = Callable[[LoggerAndParams[P, R]], ParamsOnly[P, R]]
InjectionBareProvide = Callable[[LoggerAndParams[P, R]], LoggerAndParams[P, R]]

Function = (
    SelfLoggerAndParams[S, P, R]
    | SelfAndParams[S, P, R]
    | SelfAndLogger[S, R]
    | SelfOnly[S, R]
    | LoggerAndParams[P, R]
    | ParamsOnly[P, R]
    | LoggerOnly[R]
    | Empty[R]
)
Decorator = (
    NoInjectionSelf[S, P, R]
    | InjectionSelf[S, P, R]
    | InjectionSelfProvide[S, P, R]
    | NoInjectionBare[P, R]
    | InjectionBare[P, R]
    | InjectionBareProvide[P, R]
)
```

And then I have overrides on the top level decorator method and the second level for the various scenarios. The top level overrides are somewhat messy as the only way I could figure out how to do them was to add some typing only parameters to the decorator so the overrides had something to differentiate on. Below is an example of some of the top level overrides and then a complete set of the inner overrides. It unfortunately appeared that I had to explicitly spell out every permutation of parameters provided or left to defaults.

```python
    @overload
    def decorator(
        self,
    ) -> NoInjectionSelf[S, P, R]:
        ...

    @overload
    def decorator(
        self,
        _typing_requested: Literal[False],
    ) -> NoInjectionSelf[S, P, R]:
        ...
    @overload
    def decorator(
        self,
        _typing_self: Literal[True],
        _typing_requested: Literal[False],
    ) -> NoInjectionSelf[S, P, R]:
        ...

    ### Instance False
    @overload
    def decorator(
        self,
        _typing_self: Literal[False],
    ) -> NoInjectionBare[P, R]:
        ...
    def decorator(
        self,
        _typing_self: bool = True,
        _typing_requested: bool = False,
    ) -> Decorator[S, P, R]:
```

```python
        @overload
        def decorate(
            function: SelfLoggerAndParams[S, P, R],
        ) -> SelfAndParams[S, P, R] | SelfLoggerAndParams[S, P, R]:
            ...

        @overload
        def decorate(
            function: LoggerAndParams[P, R],
        ) -> ParamsOnly[P, R] | LoggerAndParams[P, R]:
            ...

        @overload
        def decorate(
            function: SelfAndParams[S, P, R],
        ) -> SelfAndParams[S, P, R]:
            ...

        @overload
        def decorate(
            function: ParamsOnly[P, R],
        ) -> ParamsOnly[P, R] | Empty[R]:
            ...

        def decorate(function: Function[S, P, R]) -> Function[S, P, R]:
```

**Reply by erictraut** [reply]
The design of `ParamSpec` unfortunately doesn't offer a good solution for handling both instance and non-instance methods. That leads to verbose solutions like in your code sample above.
